oxidation/crates/parsec_client_types/src/local_manifest.rs

```diff
@@ -111,6 +183,99 @@ impl From<LocalFileManifest> for LocalFileManifestData {
 
 impl_local_manifest_dump_load!(LocalFileManifest);
 
+impl LocalFileManifest {
+    pub fn get_chunks(&self, block: usize) -> Vec<Chunk> {
+        self.blocks.get(block).cloned().unwrap_or_default()
+    }
+
+    pub fn is_reshaped(&self) -> bool {
+        for chunks in self.blocks.iter() {
+            if chunks.len() != 1 || !chunks[0].is_block() {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn assert_integrity(&self) {
+        let mut current = 0;
+        for (i, chunks) in self.blocks.iter().enumerate() {
+            assert_eq!(i as u64 * *self.blocksize, current);
+            assert!(!chunks.is_empty());
+            for chunk in chunks {
+                assert_eq!(chunk.start, current);
+                assert!(chunk.start < chunk.stop.into());
+                assert!(chunk.raw_offset <= chunk.start);
+                assert!(u64::from(chunk.stop) <= chunk.raw_offset + u64::from(chunk.raw_size));
+                current = chunk.stop.into()
+            }
+        }
+        assert_eq!(current, self.size);
+    }
+
+    pub fn from_remote(remote: FileManifest) -> Result<Self, &'static str> {
+        let base = remote.clone();
+        let blocks = remote
+            .blocks
+            .into_iter()
+            .map(Chunk::from_block_access)
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .map(|b| vec![b])
+            .collect();
+
+        Ok(Self {
+            base,
+            need_sync: false,
+            updated: remote.updated,
+            size: remote.size,
+            blocksize: remote.blocksize,
+            blocks,
+        })
+    }
+
+    pub fn to_remote(
+        &self,
+        author: DeviceID,
+        timestamp: DateTime,
+    ) -> Result<FileManifest, &'static str> {
+        self.assert_integrity();
+        assert!(self.is_reshaped());
+
+        let blocks = self
+            .blocks
+            .iter()
+            .map(|chunks| chunks[0].get_block_access())
+            .collect::<Result<Vec<_>, _>>()?
```

Comment:
it seems the collect cannot fail as long as `self.is_reshaped() == true`

That would mean that we could use a `map_err` to return `Err("Need reshape")`, and remove the `is_reshaped` call entirely \o/